Fixing MCGRIDINFO definition

vladimir-krestov opened the pull request:
Fixes #2912 
Related issue #2475

`MonthCalendarAccessibleObject` gets incorrect `SYSTEMTIME `info when sending Windows messages due to incorrect `MCGRIDINFO `definition.

## Proposed changes

<!-- - Guard variables against incorrect values to avoid exceptions -->

- Edit MCGRIDINFO to avoid SYSTEMTIME value errors

## Customer Impact

- A user can use a correct MonthCalendar without exceptions

## Regression? 

- Yes, from PR #2090

## Risk

- Minimal

## Screenshots 

### Before
- Incorrect SYSTEMTIME value
![image](https://user-images.githubusercontent.com/49272759/75328598-4fc37a80-588f-11ea-8d66-b3ae2c0d2c6a.png)


### After

- Correct SYSTEMTIME value
![image](https://user-images.githubusercontent.com/49272759/75327702-e8f19180-588d-11ea-90ff-13f42117c7f2.png)


## Test methodology

- CTI
- Manual UI testing
- Unit testing

## Accessibility testing 
- Using Inspect and Narrator tools
 

## Test environment(s) 

- .NET Core SDK: 5.0.0-alpha.1.20072.3
- Microsoft Windows [Version 10.0.18363.592]



###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2911)

vladimir-krestov commented on src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.SendMessageW.cs:
```diff
@@ -4,11 +4,19 @@
+            WM Msg,
+            IntPtr wParam,
+            [In, Out] ref MCGRIDINFO gridInfo);
+
```
I used this method because the existing method returns incorrect `SYSTEMTIME` value in `MCGRIDINFO`.
@JeremyKuhne is it correct? Or do I need to use the existing `SendMessageW` method with `IntPtr` parameter? If so, could you please say how?

weltkante replied:
Whats the previous caller/callee that returns incorrect values? Can't find any changed code in the PR yet (assuming not pushed yet because WIP). We might want to review the existing code for interop mistakes, adding a workaround may hide a real bug.

vladimir-krestov replied:
@weltkante, please reproduce this bug:
- Use the latest master branch
- Run an Core 5.0 app with MonthCalendar ("Calendar" app exists in the project:  tests/integration/WinformsControlsTest)
- Chose some date
- Catch the exception in DateTimePicker.SysTimeToDateTime method
- Look at values of SYSTEMTIME variable - incorrect
![image](https://user-images.githubusercontent.com/49272759/75336905-0b3edb80-589d-11ea-8109-cc4c3de5a5be.png)

This is due to incorrect getting `MCGRIDINFO` value in `MonthCalendarAccessibleObject.GetCalendarGridInfo` method 
(line 503:)
```cs
return User32.SendMessageW(_owner, (User32.WM)MCM.GETCALENDARGRIDINFO, IntPtr.Zero, ref gridInfo) != IntPtr.Zero;
```

Here is used `Interop.User32.SendMessageW` external method from `user32.dll` (`Interop.SendMessageW.cs` file, line 13 in the master branch). This method returns `lParam` with incorrect `SYSTEMTIME` info.

Please check and let me know if you need more info.

RussKie replied:
Some changes were made in b1366175621fe9af62da12799f9e3bf01a130f04, but they look sensible.
Try resetting to the previous commit (e.g. c5ebf0de0c9388d0c7ea183eac17970bdfe69f66) and see if it worked correctly.

RussKie replied:
I can confirm it worked as far back as 6a5188f3f0cb30c6adbebd12024ff1907cea01b0 (I'm on the TaskDialog branch).

vladimir-krestov replied:
Thanks for the info. 
I confirm `ManthCalendar` was broken in PR #2630. Commit c5ebf0de0c9388d0c7ea183eac17970bdfe69f66 worked well but then line 507 was changed in b1366175621fe9af62da12799f9e3bf01a130f04 commit.


____________________________
```cs
return _owner.SendMessage((int)ComCtl32.MCM.GETCALENDARGRIDINFO, 0, ref gridInfo) != IntPtr.Zero;
```
changed to 
```cs
return User32.SendMessageW(_owner, (User32.WindowMessage)MCM.GETCALENDARGRIDINFO, IntPtr.Zero, ref gridInfo) != IntPtr.Zero;
```
-----------------------------------
Actually, I restored this method with its signature in `Interop.SendMessageW.cs`

So `SendMessageW` with `IntPtr` works incorrectly for `MonthCalendar`.

vladimir-krestov replied:
@weltkante you are awesome, it works well. Thank you!

vladimir-krestov replied:
I made changes

RussKie replied:
...and we can test this too!
https://github.com/dotnet/winforms/pull/2851/files?file-filters%5B%5D=.cs#diff-77c12611fe3f980b709636a532273fc3

weltkante replied:
If you write a test make sure you test both via `sizeof` and pointer arithmetic (as in above link) as well as `Marshal.SizeOf`/`Marshal.OffsetOf`. `sizeof` and pointer arithmetic tests blitting while Marshal API tests classic marshaling. As described above they can behave differently for this struct and it'd be good to have test coverage for both to make sure this doesn't regress again.

hughbe commented on src/System.Windows.Forms.Primitives/src/Interop/User32/Interop.SendMessageW.cs:
```diff
@@ -97,5 +105,14 @@ internal static partial class User32
+        public static IntPtr SendMessageW(
+            IHandle hWnd,
+            WM Msg,
+            IntPtr wParam,
```
I don’t like comctl structs infecting this file

weltkante replied:
See discussion above, this was a workaround and not fixing the bug. It can be reverted. It is not necessary to have a separate `SendMessageW` here.

vladimir-krestov replied:
Please see https://github.com/dotnet/winforms/pull/2911#discussion_r384452969
I check, it really works. I'll remove this method.

vladimir-krestov commented:
@RussKie, @weltkante, @hughbe, please confirm and I'll give these changes to testing.

JeremyKuhne commented on src/System.Windows.Forms.Primitives/src/Interop/ComCtl32/Interop.MCGRIDINFO.cs:
```diff
@@ -22,7 +22,7 @@ public unsafe struct MCGRIDINFO
             public int iRow;
             public int iCol;
-            public bool bSelected;
+            public BOOL bSelected;
```
This is good, but `cchName` below is also wrong. `size_t` should be `UIntPtr`. It is `uint` on x32 and `ulong` on x64.

vladimir-krestov replied:
@JeremyKuhne, are you mean MCGRIDINFO should be like this?
![image](https://user-images.githubusercontent.com/49272759/75455390-73b7b680-598a-11ea-9945-af4c0b0eb10c.png)

JeremyKuhne replied:
Yes, that is correct.

vladimir-krestov replied:
Thank you!

weltkante commented:
The interop fix would probably also fix #2912

vladimir-krestov commented:
Adding unit tests...

vladimir-krestov commented:
Added **`sizeof`** unit tests like [PRINTDLGWTests](https://github.com/dotnet/winforms/blob/master/src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs)
Added **`Marshal`** unit tests as @weltkante said https://github.com/dotnet/winforms/pull/2911#discussion_r384964092

/cc @RussKie

vladimir-krestov commented:
The changes from the first commit have been added in PR #2975

vladimir-krestov commented:
Testing the fix for #2912 and #2475 in progress...

RussKie commented on src/System.Windows.Forms.Primitives/tests/Interop/ComCtl32/MCGRIDINFOTests.cs:
```diff
@@ -0,0 +1,116 @@
+        public unsafe void MCGRIDINFO_x32_Marshal_Size()
+        {
+            Assert.Equal(84, Marshal.SizeOf<MCGRIDINFO>());
+        }
```
Just FYI: there is no pressing need to test the classic marshaller, since it shouldn't be used it for blittable structs.

vladimir-krestov replied:
Do you mean I need to remove this?
Can I keep it here?

vladimir-krestov replied:
https://github.com/dotnet/winforms/pull/2911#discussion_r384964092

weltkante replied:
since this struct actually contains bools and this particular layout difference was cause of a regression I think there is some value to test both in order to prevent future regressions

RussKie commented on src/System.Windows.Forms.Primitives/tests/Interop/ComCtl32/MCGRIDINFOTests.cs:
```diff
@@ -0,0 +1,116 @@
+    public class MCGRIDINFOTests : IClassFixture<ThreadExceptionFixture>
+    {
+        public static bool Is32bit => IntPtr.Size == 4;
+        public static bool Is64bit => IntPtr.Size == 8;
```
Please move these out to src/Common/tests/, and update https://github.com/dotnet/winforms/blob/master/src/System.Windows.Forms.Primitives/tests/Interop/Comdlg32/PRINTDLGWTests.cs as well.

vladimir-krestov replied:
@RussKie, I moved these to the new **`ArchitectureDetection`** class and updated `MCGRIDINFOTests` and `PRINTDLGWTests` tests. Please check if I made it correctly.

RussKie commented:
👍
________________________________
From: Tobias Käs <notifications@github.com>
Sent: Monday, 16 March 2020 6:58 PM
To: dotnet/winforms <winforms@noreply.github.com>
Cc: Igor Velikorossov <Igor.Velikorossov@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [dotnet/winforms] WIP: Fixing MCGRIDINFO definition (#2911)


@weltkante commented on this pull request.

________________________________

In src/System.Windows.Forms.Primitives/tests/Interop/ComCtl32/MCGRIDINFOTests.cs<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fwinforms%2Fpull%2F2911%23discussion_r392838468&data=02%7C01%7Cigor.velikorossov%40microsoft.com%7C79f936f94d6b43fad97108d7c97fdbbe%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637199423297178871&sdata=ME4ueMMvdfuHbJ2ihwiT%2F%2BCEfRsW8IMbAsj64yLaG84%3D&reserved=0>:

> +        [ConditionalFact(nameof(Is32bit))]
+        public unsafe void MCGRIDINFO_x32_Marshal_Size()
+        {
+            Assert.Equal(84, Marshal.SizeOf<MCGRIDINFO>());
+        }


since this struct actually contains bools and this particular layout difference was cause of a regression I think there is some value to test both in order to prevent future regressions

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fwinforms%2Fpull%2F2911%23discussion_r392838468&data=02%7C01%7Cigor.velikorossov%40microsoft.com%7C79f936f94d6b43fad97108d7c97fdbbe%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637199423297188868&sdata=4DttzLyeSuWUdvEqvmvRBxPIKnGjH7eBIdbPkBcEWeM%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABBTEXQHYZIWVEEN6PVMHRLRHXL3PANCNFSM4K4AF5RA&data=02%7C01%7Cigor.velikorossov%40microsoft.com%7C79f936f94d6b43fad97108d7c97fdbbe%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637199423297188868&sdata=%2BrjNcMb%2F%2FgSmo6chd9lE%2FfDkNdyci%2BOoMDGZlDR7%2BD4%3D&reserved=0>.

vladimir-krestov commented:
Testers approved that Issue #2912 has been fixed. ✔️ 
Related Issue #2475 will be fixed by PR #2975.

vladimir-krestov commented:
PR is ready.
/cc @RussKie
